### PR TITLE
Approve automatic reviews if approved manual review matches

### PR DESCRIPTION
**Pull request by praveenkuttappan**
- Approve automatic review when manual review is approved if they match at API surface level
- Search for approved manual reviews whenever new automatic review is created and mark it as approved automatically.

### CODE REVIEW

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:487]: How long does this take? Would ever Approve click take forever when we get to a higher revision count?

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:498]: Do we care if the approved revision is not last?

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:461]: Why do we have logic in both directions? One that approves similar revisions and another to collect approvals? Seems that one should be enough.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:487]: This doesn't take long time since comparison us only against last revision of automatic review. But I would like this to run as background task instead of waiting for this one to update UI to avoid perf issue in the future.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:498]: In most cases, It shouldn't be. But I can think of a scenario where it breaks if we don't have API matching logic in both ways when I thought more about this comment. If someone creates a new revision as soon as last one is approved then it won't match. But this scenario is already handled by matching logic in both directions so system finds matching automatic review when manual review is approved so checking last revision is good enough.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:461]: I have added checks in both directions to ensure all scenarios are covered. If manual review is already approved(if manual review was not created from master), then matching auto review may not be present at approval time.  So checking for a matching approved review at time automatic review creation finds a match and mark it approved.

If manual review is created after merging it on to master then automatic review may be already present. So checking for existing reviews for same package when one is approved ensures existing automatic review is matched and approved. This step also ensures of marking other pending reviews for same package as outdated so we can block approval of two different changes for same package at same time which takes care of two ppl working on same package in parallel and creates API reviews in parallel.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs]: This matching logic is only triggered when scheduled automatic review is created as per latest change so this doesn't impact when approve button is clicked.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs]: I have made a change to look for last approved revision and compare against it.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs]: Removed the logic when approve button is clicked.

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:65]: FirstOrDefault.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:471]: Is this comment incorrect? It says `automatic reviews` but the parameter is `false`

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:484]: nit: ws

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:478]: Are we marking anything here?

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:446]: Do we need this optimization?

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:475]: `.Where(r => r.Approvers.Count() > 0)` can't return null.

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:70]: At this point, we might as well extend the other `GetReviewsAsync` with an additional argument.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:353]: Is it worth passing just the revision in? Instead of the whole review?

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:63]: You don't need to check count with the FirstOrDefault

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:422]: You don't need the `approvers.Count() > 0` check

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:421]: Approvers can never be null.

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:70]: I would like to do this in separate PR if that's fine.

**Review by pakrym — APPROVED** [review]

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:453]: Why `LastOrDefault` and not all the revisions?

Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:453]: Reason is to compare against latest revision of approved review rather than against all approved revisions which will be a lot of processing when comparing against old one unnecessarily. In most cases, automatically created master review should match with latest approved revision.

**Pull request merged**
